### DISCUSSION TITLE
New Auto-PID Feature with VW MEB (ID.3) and more

### DISCUSSION THREAD

**Discussion by cr0i** [meatpiHQ/wican-fw#185] [Q&A] [answered]
Hello,

I´m trying the new Auto-PID Feature with an Cupra Born (eq. VW ID.3), but I dont get valid data.
I installed Firmware 3.11, because 3.25 crashed my WiCAN.
I have no filters configured, under Automate I configured:
Initialisation: ATSP7;ATAT1;ATST96;
Name: soc
PID: 22028C
Expression: B4/2.5*51/46-6.4
Then I changed the protocoll to Auto-PID .

In the MQTT Broker I can see new entries:
others/wican/test/soc {"soc":-5.51304347826087,"raw":"F10A030F0201"}
others/wican/xxx/can/rx {"bus":"0","type":"rx","ts":27189,"frame":[{"id":417001738,"dlc":8,"rtr":false,"extd":true,"data":[3,127,34,49,170,170,170,170]}]}

These values are always the same, even i configure an other PID.

With Carscanner I connected to the soc:
[Carscanner-SoC_Log_20240828.txt](https://github.com/user-attachments/files/16788377/Carscanner-SoC_Log_20240828.txt)

The working MQTT topics without Auto-PID was:
{"bus":"0","type":"tx","frame":[{"id":402391163,"dlc":8,"rtr":false,"extd":true,"data":[3,34,2,140,170,170,170,170]}]}

I want to write the soc value with the Auto-PID Feature direct to the MQTT Broker of my Wallbox.
The format would be (i.e.):
others/wican/test/soc 45.34
Only the value of the soc (Float) without additional JSON Information.
Can the WiCAN write this value too?

**Comment by meatpiHQ** [comment]
@cr0i  

> Expression: B4/2.5*51/46-6.4

Where did you get this expression?

> With Carscanner I connected to the soc:
> [Carscanner-SoC_Log_20240828.txt](https://github.com/user-attachments/files/16788377/Carscanner-SoC_Log_20240828.txt)

what was the SoC when this log was recorded ?

Please try to update to this firmware version.

https://github.com/meatpiHQ/wican-fw/releases/tag/v3.27

> Can the WiCAN write this value too?

At the moment it will only do json, but this can be changed.

**Reply by cr0i** [reply]
@meatpiHQ 
> > Expression: B4/2.5*51/46-6.4
> 
> Where did you get this expression?

I´m using the solution from here:
https://github.com/DerHerrW/soc_helper
With this the SoC is calculated correct.

The same expression is used here too:
https://github.com/meatpiHQ/wican-fw/issues/145#issuecomment-2184157246

> > With Carscanner I connected to the soc:
> > [Carscanner-SoC_Log_20240828.txt](https://github.com/user-attachments/files/16788377/Carscanner-SoC_Log_20240828.txt)
> 
> what was the SoC when this log was recorded ?

I´m not sure anymore, but with this data from Carscanner:
>22028C
17FE007B046202**8C**7D

I thought 8C (marked bold above) could be Byte 4, 8C is 140 decimal:
140/2.5*51/46-6.4 = ~55.58%
This value should be OK.

But with the Auto-PID Feature I think, I got wrong data:
`"raw":"F10A030F0201"`

Is the Init-String OK?

> Please try to update to this firmware version.
> 
> https://github.com/meatpiHQ/wican-fw/releases/tag/v3.27

Every time I try to disable the car specific configuration after it has been enabled, the WiCAN Firmware crashes and it is unresponsible after reboot.
Then I must write new Firmware with USB-Connection.

I´ll try a little more and report back.

> > Can the WiCAN write this value too?
> 
> At the moment it will only do json, but this can be changed.

That would be great. 👍

**Reply by cr0i** [reply]
I think the crash occoures, when I try Custom PIDs instead of car profiles.

I´ve tried to build an car profile:
`         "car_model":"VW MEB",
         "init":"ATSP7;ATAT1;ATST96;",
         "pids":[
            {
               "pid":"22028C",
               "parameters":[
                  {
                     "name":"SoC",
                     "expression":"B4/2.5*51/46-6.4",
                     "unit":"%",
                     "class":"battery"
                  }
               ]
            }
         ]
      },
`

With this I get wrong data too, but I can´t see the raw data anymore.

The rx frame looks like this:
`{"bus":"0","type":"rx","ts":31604,"frame":[{"id":417001738,"dlc":8,"rtr":false,"extd":true,"data":[3,127,34,49,170,170,170,170]}]}`
or
`{"bus":"0","type":"rx","ts":8199,"frame":[{"id":417001729,"dlc":8,"rtr":false,"extd":true,"data":[3,127,1,34,170,170,170,170]}]}`

**Reply by meatpiHQ** [reply]
> Every time I try to disable the car specific configuration after it has been enabled, the WiCAN Firmware crashes and it is unresponsible after reboot.
> Then I must write new Firmware with USB-Connection.

OK, this issue is fixed in [v3.28](https://github.com/meatpiHQ/wican-fw/releases/latest)

Btw good job with the debugging! this makes things easier.

Before making a car profile please try to add a custom PID, it's easier to debug.

**Custom Initialisation**: ATSP7;ATST96;ATSHFC007B;ATCP17;ATCRA17FE007B;ATFCSH17FC007B;ATFCSD300000;ATFCSM1;
**PID**: 22028C1
**Expression**: ((B3/2.5) * 1.109)-6.4

Let me know if this works.

**Reply by cr0i** [reply]
@meatpiHQ 
> OK, this issue is fixed in [v3.28](https://github.com/meatpiHQ/wican-fw/releases/latest)

Even the vehicle specific profile is disabled, it tries to write the values configured before.
I had to enable and delete the values under Destination Topic and Cycle time, then it was OK.
And the wican keeps living! Thanks.

The calculated soc reports now as:
`others/wican/test/soc {"soc":55.704,"raw":"0462028CAA"}`
The real SoC in car is about 68%.

When I use:
((B4/2.5)*1.109)-6.4
It´s OK:
`others/wican/test/soc {"soc":69.012,"raw":"0462028CAA"}`

Now I would only need to write the value instead of JSON string.

Next I need estimated range and perhaps odometer.
I think for this we need another initialisation string?

**Reply by cr0i** [reply]
> When I use:
((B4/2.5)*1.109)-6.4
It´s OK:
others/wican/test/soc {"soc":69.012,"raw":"0462028CAA"}

But Carscanner reports 68,99%...

**Reply by cr0i** [reply]
Hello @meatpiHQ,

Now I´ve tried the Auto-PID Feature with multiple topics.
I want to use soc,range and odo.
When I use the Init string below in the field "Custom Initialisation", then I get the right value for one of the topics.
Then I change it for the next topic and reboot again, then the next value is correct, the others are wrong again.
I´m not sure, if the Expression for range is right, this I can test again, when my estimated range is below 256km.
But how can we get all values with right values? I thougt I could use it in the Init field, but it doesn´t works.


Name | Init | PID | Expression | Period(ms) | Type | Send_to
-- | -- | -- | -- | -- | -- | --
soc | ATSP7;ATST96;ATSHFC007B;ATCP17;ATCRA17FE007B;ATFCSH17FC007B;ATFCSD300000;ATFCSM1; | 22028C1 | B4 / 2.5 * 51 / 46 - 6.4 | 10000 | MQTT_Topic | test/soc
range | ATSP6;ATST96;ATSH710;ATCP18;ATCRA77A;ATFCSH710;ATFCSD300000;ATFCSM1; | 222AB62 | (B5 * 256) + B6 | 10000 | MQTT_Topic | test/range
odo | ATSP7;ATST96;ATSHFC0076;ATCP17;ATCRA17FE0076;ATFCSH17FC0076;ATFCSD300000;ATFCSM1; | 22295A | (B4 * 65536) + (B5 * 256)+B6 | 10000 | MQTT_Topic | test/odo

A few infos to the firmware problem (crash of the dongle and other feedback):
1. When the field "Custom Initialisation" is empty, then the dongle crashes again after reboot.
2. With new configuration, I have first to upload the vehicle file if I want to disable car specific, otherwise it is enabled again.

**Reply by meatpiHQ** [reply]
@cr0i The crash issue is resolved in the latest version https://github.com/meatpiHQ/wican-fw/releases

Where did you find the Odo and the range PID?

**Reply by cr0i** [reply]
Odo I knew from the project i´m using at the moment, the range I found with carscanner, after I recognized that elm327 Logging was hardcoded topic:

`272km:
{"bus":"0","type":"tx","ts":9708,"frame":[{"id":1808,"dlc":8,"rtr":false,"extd":false,"data":[3,34,42,182,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":9723,"frame":[{"id":1914,"dlc":8,"rtr":false,"extd":false,"data":[16,10,98,42,182,1,16,1]}]}
{"bus":"0","type":"tx","ts":9725,"frame":[{"id":1808,"dlc":8,"rtr":false,"extd":false,"data":[48,0,0,170,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":9727,"frame":[{"id":1914,"dlc":8,"rtr":false,"extd":false,"data":[33,16,0,0,0,170,170,170]}]}`
`276km:
{"bus":"0","type":"tx","ts":31574,"frame":[{"id":1808,"dlc":8,"rtr":false,"extd":false,"data":[3,34,42,182,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":31593,"frame":[{"id":1914,"dlc":8,"rtr":false,"extd":false,"data":[16,10,98,42,182,1,20,1]}]}
{"bus":"0","type":"tx","ts":31596,"frame":[{"id":1808,"dlc":8,"rtr":false,"extd":false,"data":[48,0,0,170,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":31598,"frame":[{"id":1914,"dlc":8,"rtr":false,"extd":false,"data":[33,20,0,0,0,170,170,170]}]}`
`278km:
{"bus":"0","type":"tx","ts":16477,"frame":[{"id":1808,"dlc":8,"rtr":false,"extd":false,"data":[3,34,42,182,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":16496,"frame":[{"id":1914,"dlc":8,"rtr":false,"extd":false,"data":[16,10,98,42,182,1,22,1]}]}
{"bus":"0","type":"tx","ts":16498,"frame":[{"id":1808,"dlc":8,"rtr":false,"extd":false,"data":[48,0,0,170,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":16500,"frame":[{"id":1914,"dlc":8,"rtr":false,"extd":false,"data":[33,22,0,0,0,170,170,170]}]}`
`293km:
{"bus":"0","type":"tx","ts":34203,"frame":[{"id":1808,"dlc":8,"rtr":false,"extd":false,"data":[3,34,42,182,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":34215,"frame":[{"id":1914,"dlc":8,"rtr":false,"extd":false,"data":[16,10,98,42,182,1,37,1]}]}
{"bus":"0","type":"tx","ts":34218,"frame":[{"id":1808,"dlc":8,"rtr":false,"extd":false,"data":[48,0,0,170,170,170,170,170]}]}
{"bus":"0","type":"rx","ts":34219,"frame":[{"id":1914,"dlc":8,"rtr":false,"extd":false,"data":[33,37,0,0,0,170,170,170]}]}`

**Reply by cr0i** [reply]
@meatpiHQ:

I have put 3 Live Data Sensors in Carscanner, SoC, ODO and Estimated Range.
Here you can see, that changing Init String is necassary, when requesting different Parameters:
[Carscanner_SoC+ODO+Range_20240904.txt](https://github.com/user-attachments/files/16874990/Carscanner_SoC%2BODO%2BRange_20240904.txt)

**Reply by meatpiHQ** [reply]
@cr0i I've added the profile for the id-3, please try the latest vehicle_profiles.json 

https://github.com/meatpiHQ/wican-fw/blob/main/vehicle_profiles.json

**Reply by cr0i** [reply]
> @cr0i I've added the profile for the id-3, please try the latest vehicle_profiles.json
> 
> https://github.com/meatpiHQ/wican-fw/blob/main/vehicle_profiles.json

Hello @meatpiHQ,

Thank you, I´ll test this in the evening.
Could the pid_init Strings used too with custom PIDs?
I have to publish the topics in different MQTT-Pathes, so I cannot use the vehicle profile for this.

**Reply by meatpiHQ** [reply]
The pid_init is not yet enabled in the custom PID.

**Reply by cr0i** [reply]
> @cr0i I've added the profile for the id-3, please try the latest vehicle_profiles.json
> 
> https://github.com/meatpiHQ/wican-fw/blob/main/vehicle_profiles.json

Sorry, it doesn´t work.
I only get status "online" but no ecu_status online.
No topics are puiblished or rx frames shown.

When I put the full init string from my table above, I can see all 3 values, but only one of them is calculated correct.
So I can see, the expressions must be OK.

Do I need a newer firmware for pid_init? I only have found 3.30.

**Reply by cr0i** [reply]
@meatpiHQ 
I´ve tried a little more with the vehicle profile.
For Testing I reduced it to one value. I´ve choosen Range for example.
When I use the profile like this:
`	{
      "car_model": "Cupra: Born",
      "init": "ATSP6;ATST96;ATFCSD300000;ATFCSM1;",
      "pids": [
        {
          "pid": "222AB62",
          "parameters": [
            {
              "name": "Range",
              "pid_init": "ATSH710;ATCP18;ATCRA77A;ATFCSH710;",
              "expression": "(B5*256)+B6",
              "unit": "km",
              "class": "none"
            }
          ]
        }
      ]
    }
`
I dont get valid data and only status online.

When I put the whole Init-String in the init value like this:
`	{
      "car_model": "Cupra: Born",
      "init": "ATSP6;ATST96;ATFCSD300000;ATFCSM1;ATSH710;ATCP18;ATCRA77A;ATFCSH710;",
      "pids": [
        {
          "pid": "222AB62",
          "parameters": [
            {
              "name": "Range",
              "expression": "(B5*256)+B6",
              "unit": "km",
              "class": "none"
            }
          ]
        }
      ]
    }
`
Then I get valid data.

**Reply by cr0i** [reply]
Hello @meatpiHQ 

do you have some news for me? 
I would like to use more than one value, but didn't get it working.

**Reply by meatpiHQ** [reply]
@cr0i sorry for the late reply, your vehicle_profile.json should look the this.


```
{
   "cars":[
		{
		   "car_model":"Cupra: Born",
		   "init":"ATSP6;ATST96;ATFCSD300000;ATFCSM1;",
		   "pids":[
			  {
				 "pid":"222AB62",
				 "parameters":[
					{
					   "name":"Range",
					   "pid_init":"ATSH710;ATCP18;ATCRA77A;ATFCSH710;",
					   "expression":"(B5*256)+B6",
					   "unit":"km",
					   "class":"none"
					}
				 ]
			  }
		   ]
		}
   ]
}
```

For reference this is the main vehicle profile that contain all the supported vehicles 

https://github.com/meatpiHQ/wican-fw/blob/main/vehicle_profiles.json

**Reply by cr0i** [reply]
Hello @meatpiHQ ,

with this setting I can´t get any data.
I uploaded this vehicle profile, deleted my custom PID setting and rebootet.
The Status is only "online" not "ecu_online" and no other topic is published.
When I start monitoring in the Monitor Tab in this moment, there comes no data too.

**Reply by meatpiHQ** [reply]
@cr0i Please update to version 3.45 and try using the new integration. This version uses HTTP to pull data from the device, so make sure to disable MQTT on the device before updating. Disabling MQTT prior to the update is important to avoid any potential issues during the firmware update.

- Disable MQTT
- Update to v3.45
- Install new HA integration 

https://github.com/jay-oswald/ha-wican
[wican-fw_obd_v345.zip](https://github.com/user-attachments/files/17246469/wican-fw_obd_v345.zip)

**Reply by cr0i** [reply]
Hello @meatpiHQ ,

the MQTT Feature was the reason, why I bought the WiCan.
I don't have HA, so I can't use the Integration. 

My Wallbox has an own MQTT Broker, where I want to publish the SoC, the Range and perhaps the Odometer. 
These Topics has different MQTT Pathes. 
So the Custom PIDs would be perfect for this. 
One Problem I have are only the different Initialisation Strings for these PIDs.
pid_init seems not to work like expected.

**Reply by cr0i** [reply]
I tried FW 3.45 with the vehicle_profile "Cupra: Born" above you gave me. The issue is still the same, I get no data.

1. I disabled MQTT
2. I updated FW to 3.45
3. I enabled MQTT and AutoPID again.

**Reply by meatpiHQ** [reply]
@cr0i mqtt will always be supported. You are correct, for what you're doing custom PID is the best solution.

> pid_init seems not to work like expected.

I'll fix this and send you a new firmware later to test.

**Reply by cr0i** [reply]
In the end I´ll need the option to write the value without xml like told above for the range.
For SoC (manual_soc below) it works, because this is a little different, this values must be written:
![image](https://github.com/user-attachments/assets/faf5cf82-6aae-4286-af12-7e669fd7f2cd)

**Reply by meatpiHQ** [reply]
@cr0i XML? Did you mean JSON?

**Reply by cr0i** [reply]
Yes, JSON is correct. Sorry.

**Reply by meatpiHQ** [reply]
@cr0i please update the v3.46, make sure to disable mqtt before the update.

After the update you can disable the Vehicle Specific PIDs and add everything in the custom PIDs, the pid init should work now.

[wican-fw_obd_v346.zip](https://github.com/user-attachments/files/17270228/wican-fw_obd_v346.zip)

> Yes, JSON is correct. Sorry.

Can you give me an example of the format?

**Reply by cr0i** [reply]
Now the custom PIDs are working (as JSON) for multiple topics.
The given format should be raw (in MQTT-Explorer):
![image](https://github.com/user-attachments/assets/c4e4c57d-74c5-425d-a926-6c962bb6e156)

![image](https://github.com/user-attachments/assets/be176dd6-8d25-430e-9483-a98325cf7932)

Name | Init | PID | Expression | Period(ms) | Type | Send_to
-- | -- | -- | -- | -- | -- | --
range | ATSP6;ATSH710;ATCP18;ATCRA77A;ATFCSH710; | 222AB62 | (B5*256)+B6 | 10000 | MQTT_Topic | openWB/set/vehicle/2/get/range

Perhaps you could add an checkbox after "Send_to" like "Raw Data"?
Then publish only the calculated value from expression to the topic in "Send_to". The "Name" must not be appended in this case.

P.S:
I´ve tried the pid_init in an vehicle_profile, there I now get timeouts. (Only for your info, I don´t need it in the moment)

**Reply by cr0i** [reply]
Hello @meatpiHQ ,

sometimes I saw timeouts with the odometer value.
I have found multiple types of odometer in Carscanner:

`ODO:`
`[8C.BMS] Wegstreckenzähler: (19760 km)`
`{"bus":"0","type":"tx","ts":40232,"frame":[{"id":402391163,"dlc":8,"rtr":false,"extd":true,"data":[3,34,2,189,170,170,170,170]}]}`
`{"bus":"0","type":"rx","ts":40236,"frame":[{"id":402522235,"dlc":8,"rtr":false,"extd":true,"data":[16,13,98,2,189,131,0,77]}]}`
`{"bus":"0","type":"tx","ts":40238,"frame":[{"id":402391163,"dlc":8,"rtr":false,"extd":true,"data":[48,0,0,170,170,170,170,170]}]}`
`{"bus":"0","type":"rx","ts":40240,"frame":[{"id":402522235,"dlc":8,"rtr":false,"extd":true,"data":[33,48,0,0,98,74,28,177]}]}`

`[01.ENG] Wegstreckenzähler: (19760 km)`
`{"bus":"0","type":"tx","ts":11438,"frame":[{"id":402391158,"dlc":8,"rtr":false,"extd":true,"data":[3,34,41,90,170,170,170,170]}]}`
`{"bus":"0","type":"rx","ts":11449,"frame":[{"id":402522230,"dlc":8,"rtr":false,"extd":true,"data":[6,98,41,90,**0,77,48**,170]}]}`

`[01.ENG] Zurückgelegte Gesamtstrecke: (19760.8 km)`
`{"bus":"0","type":"tx","ts":40303,"frame":[{"id":402391158,"dlc":8,"rtr":false,"extd":true,"data":[3,34,248,23,170,170,170,170]}]}`
`{"bus":"0","type":"rx","ts":40314,"frame":[{"id":402522230,"dlc":8,"rtr":false,"extd":true,"data":[16,19,98,248,23,0,0,38]}]}`
`{"bus":"0","type":"tx","ts":40317,"frame":[{"id":402391158,"dlc":8,"rtr":false,"extd":true,"data":[48,0,0,170,170,170,170,170]}]}`
`{"bus":"0","type":"rx","ts":40319,"frame":[{"id":402522230,"dlc":8,"rtr":false,"extd":true,"data":[33,7,0,3,3,232,255,255]}]}`
`{"bus":"0","type":"rx","ts":40321,"frame":[{"id":402522230,"dlc":8,"rtr":false,"extd":true,"data":[34,255,255,255,255,255,255,170]}]}`

`[51.ElDrive] Wegstreckenzähler: (19760 km)`
`{"bus":"0","type":"tx","ts":44450,"frame":[{"id":402391164,"dlc":8,"rtr":false,"extd":true,"data":[3,34,8,109,170,170,170,170]}]}`
`{"bus":"0","type":"rx","ts":44457,"frame":[{"id":402522236,"dlc":8,"rtr":false,"extd":true,"data":[7,98,8,109,0,0,77,48]}]}`

Which of this will fit best?
In the moment I used [01.ENG].
So I tried to use [08.BMS] with this values:

Name | Init | PID | Expression | Period(ms) | Type | Send_to
-- | -- | -- | -- | -- | -- | --
odometer | ATSP7;ATSHFC007B;ATCP17;ATCRA17FE007B;ATFCSH17FC007B; | 2202BD2 | [B6:B8] | 10000 | MQTT_Topic | others/wican/test/odometer

But with this the last digit (B8) is 0 instead of 48.
I think the multiple answer (B8 I thought is the first usable digit from the second rx frame) is ignored in this case?
What have I done wrong, or what expression should I use?

Or should I better use some other?

**Reply by meatpiHQ** [reply, marked as answer]
@cr0i Please try this firmware version, I've added support for the wallbox.

> ODO:
[8C.BMS] Wegstreckenzähler: (19760 km)

And try this expression: `(B7*256)+B9`

<img width="443" alt="image" src="https://github.com/user-attachments/assets/3f1cdfbc-f507-4482-817f-a27a651ede09">

[wican-fw_obd_v346-1-gbbe90bc.zip](https://github.com/user-attachments/files/17293642/wican-fw_obd_v346-1-gbbe90bc.zip)

**Reply by cr0i** [reply]
Hello @meatpiHQ ,

that´s great with this new Option everything I need in the moment works like expected.
Thanks for your work.

The new value for odometer doesn´t work.
If I put B8 or B9 in expression, then 0 is the value.
In the JSON answer I don´t see the 2nd message, so I think, this could be the problem.

The odometer should be 21380 at this moment:
`{"odometer":21248,"raw":"100D6202BDB30053"}`

**Reply by meatpiHQ** [reply]
@cr0i looks like its only receiving 1 frame try changing `ATST96;` to `ATSTFF;`

**Reply by cr0i** [reply]
Hello @meatpiHQ ,

with ATSTFF it seems to be the same.
I´ve found an Carscanner Log, where I requested this PID, could this help?
[log 12_2202BD2.txt](https://github.com/user-attachments/files/17319732/log.12_2202BD2.txt)

**Reply by meatpiHQ** [reply]
@cr0i please try this

Init: `ATSP7;ATSHFC007B;ATCP17;ATCRA17FE007B;ATFCSH17FC007B;ATFCSD300000;ATFCSM1;`

**Reply by cr0i** [reply]
Hello @meatpiHQ ,

sorry for the long delay.
I tried today again with firmware 3.48 and your infos.

With this full init string given in the init field it works.
I´m wondering wich of the init string should be in Custom Initialisation and wich to be in the Init field?
I tried without Custom Initialisation, but then it does not work at all.
So I put only "AT;" in the custom initialisation and in every value the full init string. So it works.

The Expression gives the right value now, but I think, when the odometer is above 65536 it would be wrong.
I think there has to be anything else multipicated with 65536?
The Expression is now: (B7*256)+B9

Thanks

**Reply by meatpiHQ** [reply]
@cr0i yes probably, I'm looking at the expression now and it feels a bit off. What is the response of PID 2202BD2 now?

**Reply by cr0i** [reply]
I noticed that the PID 2202BD2 has timeouts too and then the value was suddenly wrong 65534.
So I changed back to the old PID, that didn´t gave the wrong value when timed out.
I think the timeouts couldn´t be changed. It was only a try.
Sorry for the additional work I generated to you.
When it´s important for you I can go on testing with PID 2202BD2, when you want.

So long the only question I have is how I should split the Initialisation String (In Custom Init and Init Field)?

**Reply by cr0i** [reply]
Hello @meatpiHQ ,

I wanted to add a new Custom PID in my config.
In the moment I have 3 lines Custom PIDs.
When I add a new line, so I have 4 Custom PIDs, the config is empty after reboot and vehicle profiles are enabled by default.
Firmware 3.48 is used.

The new Custom PID (perhaps you want to add it to the vehicle Profile):
Max. Capacity of HV Battery:

Name | Init | PID | Expression
-- | -- | -- | --
capacity | ATSP6;ATSH710;ATCP18;ATCRA77A;ATFCSH710; | 222AB2 | [B4:B5]*50/1000

The Expression I got from ABRP, the value is the same as in carscanner, so it should be OK.

**Reply by gerhardvilsecker** [reply]
@cr0i: may I get back to your first question in this thread? You asked for [01.ENG] odometer. For me this is the most interesting odometer value because it scales to 0.1 km steps. Finally I got it working. For anyone who cares. Here is my custom_pid setting:
{
        "Name": "odometerdecimal",
        "Init": "ATCP18;ATSHDB33F1;ATSHFC0076;ATCP17;ATCRA17FE0076;ATFCSH17FC0076;",
        "PID": "22F817",
        "Expression": "(B11*65536 +B12*256 + B13)/10",
        "MinValue": "",
        "MaxValue": "",
        "Period": "5000",
        "Type": "MQTT_Topic",
        "Send_to": "wican/cupraborn/can/"odometerdecimal
      }

Thanks a lot for giving me the right entry point for my research :-)

PS: Engine ECU seems to be one of those ECUs that do only answer when car is driving. So dont expect to get answer when car is parked...

**Reply by gerhardvilsecker** [reply]
multiply sign is removed every time I copy the equation. Sorry for that. Equation is : (B11 times 2^16 + B12 times 2^8 + B13)/10

**Comment by typxxi** [comment]
I doubt that the range is done this way.

I was told the same and at the end it was completely different, when I had to do that for the e-GOLF

Why ? There were first of all comments in some documents like gross 0,4 

Then I saw in carscanner app that the charging progress steps were always 0,4% like from 60% to 60,4 and 60,8% 

From there I started the assumption that the whole SOC was covered by 250 steps from 0 to 250 each 0,4%

Next step: I started the car fully, which means the wallbox stopped charging.
Guess what: Carscanner said 96% and the value was 240.
Final step: driving the car to range 0 and checking again: 8% and 20 was the value from the control unit.

Again: this is the gross soc and I can only recommand to go for this SOC, nothing else. 

And then after all these tests I am using the gros SOC range from 8% to 96% as my SOC 0% - 100%.
cars gross 8% is my 0% and cars 96% is my 100%

Quite accurate and also easy to check. 
Many people assume 256 as full for whatever reason but the gross battery soc should show the step width in your car scanner app. If the value has a fraction go for the gross battery calculation cause I had found 2 SOC for my car. 

ATTENTION: the gross battery soc shown by car scanner pro app has a step width of 0,4% but my SOC from 8% to 96% has different step width cause 220 steps for 100% means a bigger step width. 

VW is using that in more cars, even hybrid.

**Reply by cr0i** [reply]
> I doubt that the range is done this way.
> 
> I was told the same and at the end it was completely different, when I had to do that for the e-GOLF
> 
> Why ? There were first of all comments in some documents like gross 0,4
> 
> Then I saw in carscanner app that the charging progress steps were always 0,4% like from 60% to 60,4 and 60,8%
> 
> From there I started the assumption that the whole SOC was covered by 250 steps from 0 to 250 each 0,4%
> 
> Next step: I started the car fully, which means the wallbox stopped charging. Guess what: Carscanner said 96% and the value was 240. Final step: driving the car to range 0 and checking again: 8% and 20 was the value from the control unit.
> 
> Again: this is the gross soc and I can only recommand to go for this SOC, nothing else.
> 
> And then after all these tests I am using the gros SOC range from 8% to 96% as my SOC 0% - 100%. cars gross 8% is my 0% and cars 96% is my 100%
> 
> Quite accurate and also easy to check. Many people assume 256 as full for whatever reason but the gross battery soc should show the step width in your car scanner app. If the value has a fraction go for the gross battery calculation cause I had found 2 SOC for my car.
> 
> ATTENTION: the gross battery soc shown by car scanner pro app has a step width of 0,4% but my SOC from 8% to 96% has different step width cause 220 steps for 100% means a bigger step width.
> 
> VW is using that in more cars, even hybrid.

I have multiple SoC´s in Carscanner but no one of them is the same as the SoC displayed in the car.
First of all, I need multiple topics, that can be published with AutoPID, then I can analyse, wich SoC will fit best.
In the moment the SoC read is a little bit lower than that displayed, that is OK for now.

**Comment by mattberlinHA** [comment]
Hi guys,

I am using this great Auto PID feature in order to transmit the SoC and the mileage to the Wallbox.

On the other hand, using the WiCAN interface in ELM327 mode allows you to comfortably access the car's data with a smart phone app, such as Car Scanner via WLAN.

Is there a way to combine the 2 features Auto PID and ELM327 at the same time?

**Comment by SHentschke** [comment]
Hi @meatpiHQ ,

I have the WiCAN-OBD-PRO with software version 4.48. Are there any new values ​​available for the VW ID. family?

I need:
- Speed ​​(the speed of standard PIDs sometimes shows 34 km/h even when the vehicle is stationary)
- Power (charging and discharging power)
- Ambient temperature
- Interior temperature
- Battery temperature
- GPS data (latitude/longitude)
- Humidity
- Air pressure
- Charge cycles

The most important values ​​would be speed, power, and GPS data. But anything else would be great too.
Thanks in advance.

Regards,
Stefan

**Reply by cr0i** [reply]
I have some additional data, but not the ones important for you.
I can create a PR for them.

**Reply by SHentschke** [reply]
Hello, thank you for the quick reply. I'm grateful for any additional PID.
But what do you mean by "I can create a PR for them"?
